File: transaction-analyzer/java/src/main/java/com/example/demo/InputStreamTransactionLoader.java
package com.example.demo;

import com.example.demo.Transaction;
import com.example.demo.TransactionLoader;
import com.example.demo.TransactionType;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.stream.Collectors;

public final class InputStreamTransactionLoader implements TransactionLoader {
    final InputStream source;

    public InputStreamTransactionLoader(InputStream source) {
        this.source = source;
    }

    @Override
    public List<Transaction> load() throws IOException {
        try (var reader = new BufferedReader(new InputStreamReader(this.source))) {
            return reader.lines().skip(1).map(this::buildTransaction).collect(Collectors.toList());
        }
    }

    private Transaction buildTransaction(String line) {
        System.out.println("reading line:" + line);
        var fields = line.split(",");
        System.out.println("fields: " + fields.length);
        return new Transaction(
                fields[0].trim(),
                LocalDateTime.parse(fields[1].trim(), DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm:ss")),
                new BigDecimal(fields[2].trim()),
                fields[3].trim(),
                TransactionType.valueOf(fields[4].trim()),
                fields.length == 6 ? fields[5].trim() : null
        );
    }
}

File: kotlin/src/main/kotlin/InputStreamTransactionLoader.kt
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.math.BigDecimal
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import kotlin.streams.toList

interface TransactionLoader {
    fun load(): List<Transaction>
}

internal class InputStreamTransactionLoader(val input: InputStream) : TransactionLoader {
    override fun load(): List<Transaction> {
        val reader = BufferedReader(InputStreamReader(input))
        return reader.lines().skip(1).map { buildTransaction(it) }.toList()
    }

    private fun buildTransaction(line: String): Transaction {
        println("build transaction from line: $line")
        val fields = line.split(",")
        return Transaction(
            id = fields[0].trim(),
            transactedAt = LocalDateTime.parse(
                fields[1].trim(),
                DateTimeFormatter.ofPattern(Constants.defaultDateTimeFormatter)
            ),
            amount = BigDecimal(fields[2].trim()),
            merchantName = fields[3].trim(),
            type = TransactionType.valueOf(fields[4].trim()),
            relatedTransactionId = fields[5].trim()
        )
    }
}